client/file-manager/src/rocksdb.rs

```diff
@@ -0,0 +1,1023 @@
+        let mut root = self.root;
+        let db = self.as_hash_db_mut();
+        let trie_root_key = root;
+        let mut trie = TrieDBMutBuilder::<T>::from_existing(db, &mut root).build();
+
+        for chunk_id in 0..chunk_count {
+            trie.remove(&ChunkId::new(chunk_id as u64).as_trie_key())
+                .map_err(|e| {
+                    error!(target: LOG_TARGET, "Failed to delete chunk from RocksDb: {}", e);
+                    FileStorageWriteError::FailedToDeleteChunk
+                })?;
+        }
+
+        // Remove the root from the trie.
+        trie.remove(trie_root_key.as_ref()).map_err(|e| {
+            error!(target: LOG_TARGET, "Failed to delete root from RocksDb: {}", e);
+            FileStorageWriteError::FailedToDeleteChunk
+        })?;
+
+        let new_root = *trie.root();
+
+        drop(trie);
+
+        // TODO: improve error handling
+        // Commit the changes to disk.
+        self.commit(trie_root_key).map_err(|e| {
+            error!(target: LOG_TARGET, "Failed to commit changes to persistent storage: {}", e);
+            FileStorageWriteError::FailedToPersistChanges
+        })?;
+
+        // Set new internal root (empty trie root)
+        self.root = new_root;
+
+        Ok(())
+    }
+}
+
+impl<T: TrieLayout + Send + Sync> AsHashDB<HashT<T>, DBValue> for RocksDbFileDataTrie<T>
+where
+    HasherOutT<T>: TryFrom<[u8; H_LENGTH]>,
+{
+    fn as_hash_db<'b>(&'b self) -> &'b (dyn HashDB<HashT<T>, DBValue> + 'b) {
+        self
+    }
+    fn as_hash_db_mut<'b>(&'b mut self) -> &'b mut (dyn HashDB<HashT<T>, DBValue> + 'b) {
+        &mut *self
+    }
+}
+
+impl<T: TrieLayout + Send + Sync> hash_db::HashDB<HashT<T>, DBValue> for RocksDbFileDataTrie<T>
+where
+    HasherOutT<T>: TryFrom<[u8; H_LENGTH]>,
+{
+    fn get(&self, key: &HasherOutT<T>, prefix: Prefix) -> Option<DBValue> {
+        HashDB::get(&self.overlay, key, prefix).or_else(|| {
+            self.storage
+                .read()
+                .expect("Failed to acquire read lock")
+                .get(key, prefix)
+                .unwrap_or_else(|e| {
+                    warn!(target: LOG_TARGET, "Failed to read from DB: {}", e);
+                    None
+                })
+        })
+    }
+
+    fn contains(&self, key: &HasherOutT<T>, prefix: Prefix) -> bool {
+        HashDB::contains(&self.overlay, key, prefix)
+    }
+
+    fn insert(&mut self, prefix: Prefix, value: &[u8]) -> HasherOutT<T> {
+        HashDB::insert(&mut self.overlay, prefix, value)
+    }
+
+    fn emplace(&mut self, key: HasherOutT<T>, prefix: Prefix, value: DBValue) {
+        HashDB::emplace(&mut self.overlay, key, prefix, value)
+    }
+
+    fn remove(&mut self, key: &HasherOutT<T>, prefix: Prefix) {
+        HashDB::remove(&mut self.overlay, key, prefix)
+    }
+}
+
+pub struct RocksDbFileStorage<T: TrieLayout + 'static>
+where
+    HasherOutT<T>: TryFrom<[u8; H_LENGTH]>,
+{
+    storage: Arc<RwLock<dyn Backend<T>>>,
```

Comment:
There is already a storage field on the `FileDataTrie` struct, can you add a TODO here to remove this. This struct doesn't do anything besides passing storage to the file trie struct instances. We can switch this for some lazy loaded global storage that can be access by all the file tries.